### ISSUE TITLE
Investigate AWS FXs for Lustre

### ISSUE THREAD

**Issue by tomcrane** [dlcs/protagonist#79]
See https://github.com/dlcs/protagonist/blob/master/docs/rfcs/002-storage-and-orchestration.md#amazon-fsx-for-lustre

This could be done in tandem with image server work in https://github.com/dlcs/protagonist/issues/22 

Each and every Cantaloupe instance has one or more Lustre fs mounted.
Each and every Cantaloupe instance can read from S3 or Lustre, it depends on what request Orchestrator forwards. Or, separate traffic as in #30. 

Orchestrator would then become something that doesn't actually do orchestration - doesn't move files itself - but routes and proxies with knowledge of what orchestration Lustre-S3 is now doing.

A DLCS has at least one Lustre FS available. This is its internal one, which fronts the DLCS' own S3 storage.

This replaces use of EFS, local EBS on containers, everything scratch-like, as well as catering for any image registered that isn't from a new type of optimised origin, the "customer" origin:

Wellcome has a second Lustre mounted, as read only - this is the new type of optimised origin. It is a Lustre-exposure of the storage service bucket. You can't use this one instead of EFS/EBS for scratch, because it's not yours! And it's read only.

This might be wasteful though, having two for Wellcome; Lustre is 1.2 TB minimum.

So...

A dlc.services approach could have one Lustre for everything, 1.2 TB min. No other optimised origins. Used for scratch ops and own storage.
Wellcome could have a [multi-attach](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-volumes-multi.html) EBS instead of Lustre for the first type, because DLCS only needs it for scratch, not scratch AND storage.

Both of these would need to be scavenged, for scratch-related files.
In the Wellcome scenario, the "external" Lustre would also need to be scavenged to keep the size down (see the RFC) but perhaps this is not for the DLCS to do - it doesn't know who else is using the file system. 

Types of origin the DLCs can handle:

1) A mounted filesystem (on image servers). Maybe a multi-attach file system.
2) A mounted filesystem (on image servers) that fronts an S3 bucket, where the DLCS can access it through either, and will only use file system access for speed-critical operations. This can be the DLCS' own bucket, or a customers. This would need either two values for origin, or a transformation from the existing S3 origin to its Lustre file path, which it can do because it knows the bucket is "optimised".
3) An S3 bucket that the DLCS has read access to (anywhere). DLCS no longer orchestrates from here at runtime. It either copies to its own Lustre-fronted S3 (file becomes duplicated in DLCS) or it is restricted to S3-only ops on the file (via Cantaloupe).
4) Any other reachable location (HTTP, FTP etc)

Some monthly [pricing](https://aws.amazon.com/fsx/lustre/pricing/) for 1.2 TB

HDD - 40 MB/s/TiB baseline, up to 1.3 GB/s/TiB burst - $99.60
SDD - 100 MB/s/TiB baseline, up to 1.3 GB/s/TiB burst - $250.80

Using the first of these as scratch in the Wellcome case isn't too bad and keeps the number of different storage mechanisms down.

**Comment by stephenwf** [comment]
Would a theoretical update to the DLCS allow for a local-only variation - where we just have _all_ of the files in a mount? It might be more cost-effective for sets of images less than 1.2TB

**Comment by tomcrane** [comment]
Yes definitely. The multi-attach feature means you could run a DLCS with a 1TB or 500GB shared volume (for example), self-contained.

If Lustre and a multi-attached volume both look like regular file systems then it should be OK.

**Comment by stephenwf** [comment]
And the running of `lfs hsm_release [ ... ]` would be behind config or something?

**Comment by donaldgray** [comment]
For implementation - we don't want to have to manage multiple different branching Orchestrator implementations.

Have either a LustreFXs implementation, or have an S3 + disk approach that requires orchestration of files. In the former case the Orchestrator component acts more like a pure reverse-proxy.

The preferences is to have a single Orchestrator instance running within a single AZ, excessive load would require the use of Lustre to avoid the complexity of managing orchestration file status across various instances, and the potential for stores to get out of sync.
